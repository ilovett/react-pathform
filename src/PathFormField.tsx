import React from 'react';
import {
  PathFormPath,
  PathFormPublishOptions,
  PathFormStoreMeta,
  PathFormValidation,
  toDotPath,
  usePathForm,
  usePathFormDotPath,
  usePathFormValue,
} from '.';

export interface PathFormInputProps {
  name: string;
  value: any;
  onChange: (event?: any, value?: any) => any;
  onBlur: (event?: any) => any;
}

export interface PathFormFieldRenderProps {
  inputProps: PathFormInputProps;
  meta: PathFormStoreMeta;
  renders: number;
}

export interface PathFormFieldProps {
  path: PathFormPath;
  defaultValue: any; // TODO generics?
  render: (props: PathFormFieldRenderProps) => React.ReactElement; // TODO React.ReactNode vs ReactElement ???
  renderEmpty?: (props: PathFormFieldRenderProps) => React.ReactElement; // TODO React.ReactNode vs ReactElement ???
  validations?: Array<PathFormValidation>;
  publish?: PathFormPublishOptions;
}

export const PathFormField: React.FC<PathFormFieldProps> = ({ path, render, defaultValue, validations, publish }) => {
  const name = usePathFormDotPath(path);
  const [value, meta, renders] = usePathFormValue(path, defaultValue); // TODO defaultValue needed?
  const { setValue, setMeta, clearError, watchers } = usePathForm();

  const onChange = React.useCallback(
    (event: any) => {
      // TODO handle all sorts of values
      const value = event?.target?.value ?? event;
      setValue(path, value);
      // setMeta(path, { dirty: true });

      if (publish?.path) {
        // TODO this value is wrong and should just be publish(publish?.path) -- if no value given, get the store value as-is
        watchers.current.emit(toDotPath(publish?.path), value);
      }
    },
    // ignore `watchers`
    // ignore `publish` TODO eventually this should be serialized
    // eslint-disable-next-line
    [path, setValue, setMeta]
  );

  // TODO onBlur, ref
  const onBlur = React.useCallback(
    (event: any) => {
      // TODO handle all sorts of values
      const blurValue = event?.target?.value ?? event;

      // if the field has an error, and our blur value is different than the error value, clear the error
      if (Boolean(meta?.error) && blurValue !== meta?.error?.value) {
        clearError(path);
      }

      if (meta.touched !== true) {
        setMeta(path, { touched: true });
      }
    },
    [path, clearError, setMeta, meta]
  );

  // TODO reusable hook for usePathFormValidation -> useEffect [validations]
  React.useEffect(() => {
    // TODO handle multiple fields for same path
    if (validations) {
      setMeta(path, { validations });
    }

    // unmounted the field at path
    return () => {
      // remove validations on the given path -- assuming 1 mount per path...
      // consider multiple mounts to the same store value...
      // since we're removing lets just not "remove" validations now
      // since the whole store item is getting removed...
      // TODO this should not throw errors
      if (validations) {
        // debugger;
        // setMeta(path, { validations: null });
      }
    };
    // eslint-disable-next-line
  }, []);

  return render({
    inputProps: {
      name,
      value,
      onChange,
      onBlur,
    },
    meta,
    renders,
  });
};
