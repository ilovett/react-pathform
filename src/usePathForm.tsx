import React from 'react';
import {
  set,
  get,
  typeOf,
  toDotPath,
  toStorePath,
  createStoreItem,
  arrayMove,
  arrayRemove,
  createStore,
  eventEmitter,
  parseStore,
} from '.';
import { flattenStore, PathFormStoreItemFlat } from './storeUtils';
import { noDifference } from './utils';

export type PathFormValuePrimitive = string | number | boolean | null;

export type PathFormState = {
  store: PathFormStore;
  dirtyUuids: string[];
  errors: PathFormStoreItemFlat[];
  defaultValues: any;
};

export type PathFormPath = Array<string | number>; // | string;

export type PathFormError = {
  type: string;
  message: React.ReactNode;
  value?: any;
};

export type PathFormErrorOptions = {
  publish?: boolean;
};

export type PathFormStoreMeta = {
  uuid: string;
  dirty: boolean;
  touched: boolean;
  error: null | PathFormError;
  validations: null | Array<PathFormValidation>;
  defaultValue: any;
};

// TODO omit error?
export type PathFormStoreSetMeta = Partial<Omit<PathFormStoreMeta, 'uuid'>>;

export type PathFormValidation =
  | { type: 'required'; message: string }
  | { type: 'minLength' | 'maxLength' | 'min' | 'max'; value: number; message: string }
  | { type: 'regex'; value: RegExp; message: string }
  | { type: 'custom'; value: (value: any, store?: any) => boolean; message: string };

export type PathFormValueType = 'primitive' | 'object' | 'array';

export type PathFormStorePrimitive = {
  type: 'primitive';
  meta: PathFormStoreMeta;
  value: PathFormValuePrimitive;
};

export type PathFormStoreObject = {
  type: 'object';
  meta: PathFormStoreMeta;
  value: {
    [key: string]: PathFormStoreItem;
  };
};

export type PathFormStoreArray = {
  type: 'array';
  meta: PathFormStoreMeta & { defaultFieldUuids: string[] };
  value: Array<PathFormStoreItem>;
};

export type PathFormStoreItem = PathFormStorePrimitive | PathFormStoreObject | PathFormStoreArray;

export type PathFormStoreInput = {
  [key: string]: any;
};

export type PathFormStore = {
  [key: string]: PathFormStoreItem;
};

export type PathFormResetOptions = {
  defaultValues?: any;
};

type PathFormWatcher = Readonly<{
  on: (name: string, fn: (data?: any) => any) => any;
  emit: (name: string, data: any) => any;
}>;

type PathFormStateContextInitialized = {
  state: null;
  watchers: null;
};

type PathFormStateContext = {
  state: React.MutableRefObject<PathFormState>;
  watchers: React.MutableRefObject<PathFormWatcher>;
  array: PathFormArrayUtils;
  getValues: () => any;
  setValue: (path: PathFormPath, value: any) => any;
  setMeta: (path: PathFormPath, meta: PathFormStoreSetMeta) => any;
  addError: (path: PathFormPath, error: PathFormError) => any;
  clearError: (path: PathFormPath) => any;
  clearErrors: (paths?: PathFormPath[]) => any;
  forEachStoreItem: (callback: (item: PathFormStoreItemFlat) => any) => any;
  validate: (path: PathFormPath) => any;
  validateStore: () => any; // TODO throws?
  isDirty: () => boolean;
  reset: (options?: PathFormResetOptions) => any;
};

export type PathFormArrayUtils = {
  append: (path: PathFormPath, item: any) => any;
  prepend: (path: PathFormPath, item: any) => any;
  remove: (path: PathFormPath, index: number) => any;
  move: (path: PathFormPath, fromIndex: number, toIndex: number) => any;
};

export interface PathFormPublishOptions {
  path?: PathFormPath;
  // paths?: Array<PathFormPath?; // TODO publish to multiple paths
  // up?: number; // TODO publish parent / up N levels
  // down?: number; // TODO publish child / down N levels
}

// eslint-disable-next-line
const PathFormStateContext = React.createContext<PathFormStateContextInitialized | PathFormStateContext>({
  state: null,
  watchers: null,
});

PathFormStateContext.displayName = 'PathFormStateContext';

export interface PathFormProviderProps {
  initialRenderValues?: any; // TODO generics or something
  children: React.ReactNode;
}

export const PathFormProvider: React.FC<PathFormProviderProps> = ({ children, initialRenderValues }) => {
  const state = React.useRef<PathFormState>({
    store: createStore(initialRenderValues),
    dirtyUuids: [],
    errors: [],
    defaultValues: initialRenderValues,
  });
  const watchers = React.useRef(eventEmitter());

  const getValues = () => {
    return parseStore(state.current.store);
  };

  const validateStore = () => {
    const flattenedStoreItems = flattenStore(state.current.store);

    // validate at every path
    flattenedStoreItems.forEach(({ path }) => {
      validate(path);
    });

    // look for at least one error
    const errors = flattenedStoreItems.filter(({ storeItem }) => {
      return !!storeItem.meta.error;
    });

    // throw to stop submission
    if (errors.length) {
      throw errors;
    }
  };

  const forEachStoreItem = (callback: (item: PathFormStoreItemFlat) => any) => {
    const flattenedStoreItems = flattenStore(state.current.store);
    flattenedStoreItems.forEach(callback);
  };

  const validate = (path: PathFormPath) => {
    const dotpath = toDotPath(path); // cant use hooks inside
    const storePath = toStorePath(path);
    const storeItem = get(state.current.store, storePath) as PathFormStoreItem;

    // validate that the target store item exists
    if (storeItem === undefined) {
      throw new Error(`The target path "${dotpath}" does not exist.`);
    }

    if (storeItem.meta.validations) {
      storeItem.meta.validations.forEach((validation) => {
        // do nothing if the store item already has a validation error
        // TODO this may eventually get eliminated once a field can have multiple errors
        if (storeItem.meta.error) {
          return;
        }

        if (validation.type === 'custom') {
          if (!(validation.value && validation.value(storeItem.value, getValues()))) {
            addError(path, validation);
          }
        } else if (storeItem.type === 'primitive') {
          if (validation.type === 'required') {
            if (storeItem.value === null || (typeof storeItem.value === 'string' && storeItem.value === '')) {
              addError(path, validation);
            }
          } else if (validation.type === 'min') {
            if (Number(storeItem.value) < validation.value) {
              addError(path, validation);
            }
          } else if (validation.type === 'max') {
            if (Number(storeItem.value) > validation.value) {
              addError(path, validation);
            }
          } else if (validation.type === 'minLength') {
            if (typeof storeItem.value === 'string' && storeItem.value.length < validation.value) {
              addError(path, validation);
            }
          } else if (validation.type === 'maxLength') {
            if (typeof storeItem.value === 'string' && storeItem.value.length > validation.value) {
              addError(path, validation);
            }
          } else if (validation.type === 'regex') {
            try {
              const regex = validation.value;
              if (typeof storeItem.value === 'string' && !regex.test(storeItem.value)) {
                addError(path, validation);
              }
            } catch (err) {
              // could not compile regex
            }
          }
        } else if (storeItem.type === 'array') {
          if (validation.type === 'minLength') {
            if (storeItem.value.length < validation.value) {
              addError(path, validation);
            }
          } else if (validation.type === 'maxLength') {
            if (storeItem.value.length > validation.value) {
              addError(path, validation);
            }
          }
        }
      });
    } else {
      // clearError(path);
    }
  };

  const setValue = (path: PathFormPath, value: any) => {
    const storePath = toStorePath(path);
    const storeItem = get(state.current.store, storePath) as PathFormStoreItem; // TODO default value based on value here?

    // only update store & emit to subscribers if value changed
    // TODO performance protect against non primitive types, EG: comparing `[] !== []` is true
    if (storeItem?.value !== value) {
      const newValueType = typeOf(value);

      // no store item exists at the given path, set it up!
      if (!storeItem) {
        set(state.current.store, storePath, createStoreItem(value));
      }
      // if new type is primitive, simply update the value
      else if (newValueType === 'primitive') {
        set(state.current.store, [...storePath, 'type'], 'primitive');
        set(state.current.store, [...storePath, 'value'], value);
      }
      // otherwise arrays items and object properties need to be converted to store items
      else if (newValueType === 'array' || newValueType === 'object') {
        const newStoreItem = createStoreItem(value);

        // keep the existing uuid but change the type and value
        set(state.current.store, [...storePath, 'type'], newValueType);
        set(state.current.store, [...storePath, 'value'], newStoreItem.value);
      } else {
        throw new Error('Unhandled setValue case.');
      }

      // check if field is dirty
      const dirty = storeItem?.meta.defaultValue !== value;

      // if dirty value changes, add or remove from dirtyFields
      if (!!storeItem && dirty !== storeItem.meta.dirty) {
        // determine if previously dirty
        const dirtyIndex = state.current.dirtyUuids.indexOf(storeItem.meta.uuid);

        // not dirty anymore, but previously dirty
        if (!dirty && dirtyIndex >= 0) {
          arrayRemove(state.current.dirtyUuids, dirtyIndex);
        }
        // dirty and not previously dirty
        else if (dirty && dirtyIndex < 0) {
          state.current.dirtyUuids.push(storeItem.meta.uuid);
        }

        watchers.current.emit('dirty', state.current.dirtyUuids);
      }

      set(state.current.store, [...storePath, 'meta', 'dirty'], dirty);

      watchers.current.emit(toDotPath(path), value);
    }
  };

  const setMeta = (path: PathFormPath, meta: PathFormStoreSetMeta) => {
    const dotpath = toDotPath(path); // cant use hooks inside
    const storePath = toStorePath(path);
    const storeItem = get(state.current.store, storePath);

    // validate that the target store item exists
    if (storeItem === undefined) {
      throw new Error(`The target path "${dotpath}" does not exist.`);
    }

    // mutate the target store item meta with updated values
    const existing = get(state.current.store, [...storePath, 'meta']) as PathFormStoreMeta;
    Object.assign(existing, meta);

    // notify subscribers
    watchers.current.emit(dotpath, storeItem);
  };

  const addError = (path: PathFormPath, error: PathFormError, options?: PathFormErrorOptions) => {
    const dotpath = toDotPath(path);
    const storePath = toStorePath(path);
    const storeItem = get(state.current.store, storePath); // TODO default value based on value here?

    // validate that the target store item exists
    if (storeItem === undefined) {
      throw new Error(`The target path "${dotpath}" does not exist.`);
    }

    // mutate the target store item error
    set(state.current.store, [...storePath, 'meta', 'error'], error);

    // add to errors if not already added
    if (!state.current.errors.find((flatStoreItem) => flatStoreItem.dotpath === dotpath)) {
      state.current.errors.push({ dotpath, path, storeItem });

      // TODO if (options.publish !== false)
      // notify the errors subscribers
      if (options?.publish !== false) {
        watchers.current.emit('errors', state.current.errors);
      }
    }

    // notify each the dotpath subscribers
    watchers.current.emit(dotpath, storeItem);
  };

  /**
   * Remove the error at a given path.If a specific error is given,
   * then only that error will be removed.
   */
  function clearError(path: PathFormPath, error?: PathFormError, options?: PathFormErrorOptions) {
    const dotpath = toDotPath(path);
    const storePath = toStorePath(path);
    const storeItem = get(state.current.store, storePath);

    // validate that the target store item exists
    if (storeItem === undefined) {
      throw new Error(`The target path "${dotpath}" does not exist.`);
    }

    // mutate the target store item
    set(state.current.store, [...storePath, 'meta', 'error'], null);

    // remove from the errors array if found
    const errorIndex = state.current.errors.findIndex((flatStoreItem) => flatStoreItem.dotpath === dotpath);
    if (errorIndex > -1) {
      arrayRemove(state.current.errors, errorIndex);

      // TODO if (options.publish !== false)
      // notify the errors subscribers
      if (options?.publish !== false) {
        watchers.current.emit('errors', state.current.errors);
      }
    }

    // notify subscribers
    watchers.current.emit(dotpath, storeItem);
  }

  const clearErrors = () => {
    const flattenedStoreItems = flattenStore(state.current.store);

    // TODO filter all paths if given

    // validate at every path
    flattenedStoreItems.forEach(({ path }) => clearError(path));
  };

  // helper for mutating store arrays to keep the array utils DRY
  // validate that the storeItem is an array, do some callback action, and emit
  const mutateArray = (path: PathFormPath, mutate: (targetArrayStoreItem: PathFormStoreArray) => any) => {
    const dotpath = toDotPath(path);
    const targetArrayStoreItem = get(state.current.store, toStorePath(path)) as PathFormStoreArray | undefined;

    // validate that the target value is an array
    if (targetArrayStoreItem?.type !== 'array' || !Array.isArray(targetArrayStoreItem?.value)) {
      throw new TypeError(`The target path "${dotpath}" is not an array.`);
    }

    // do the mutation
    mutate(targetArrayStoreItem);

    // dirty checks on array field
    // compare new array of fields uuid structure
    const currentFieldUuids = targetArrayStoreItem.value.map((storeItem) => storeItem.meta.uuid);
    const dirtyIndex = state.current.dirtyUuids.indexOf(targetArrayStoreItem.meta.uuid);

    // if current field state is equal to default state field uuids
    // unmark it as dirty
    if (noDifference(currentFieldUuids, targetArrayStoreItem.meta.defaultFieldUuids)) {
      targetArrayStoreItem.meta.dirty = false;

      // remove it from the dirty uuids if found
      if (dirtyIndex >= 0) {
        arrayRemove(state.current.dirtyUuids, dirtyIndex);
        watchers.current.emit('dirty', state.current.dirtyUuids);
      }
    }
    // fields look different
    else {
      targetArrayStoreItem.meta.dirty = true;

      // and not already dirty, save the array uuid to dirty fields
      if (dirtyIndex === -1) {
        state.current.dirtyUuids.push(targetArrayStoreItem.meta.uuid);
        watchers.current.emit('dirty', state.current.dirtyUuids);
      }
    }

    // notify subscribers
    watchers.current.emit(dotpath, targetArrayStoreItem);
  };

  /**
   * Returns `true` if at least one field has been modified in the form.
   */
  function isDirty() {
    return state.current.dirtyUuids.length > 0;
  }

  /**
   * Resets the form to defaultValues, or updates the defaultValues if given.
   *
   * @param options
   */
  function reset(options: PathFormResetOptions = {}) {
    // TODO currrently im being lazy and just recreating the entire store
    // which gives us new uuids -- this may lead to bugs
    // a future improvement (and/or option for reset)
    // may be to keep original uuid values in place, if possible, by merge
    const newDefaultValues = typeof options.defaultValues !== 'undefined' ? options.defaultValues : state.current.defaultValues;
    const newStore = createStore(newDefaultValues);

    state.current.store = newStore;
    state.current.defaultValues = newDefaultValues;
    state.current.dirtyUuids = [];
    state.current.errors = [];

    // notify every store item to re-render
    forEachStoreItem((flatStoreItem) => watchers.current.emit(flatStoreItem.dotpath, flatStoreItem.storeItem.value));

    // notify store meta dirty / errors
    watchers.current.emit('dirty', state.current.dirtyUuids);
    watchers.current.emit('errors', state.current.errors);
  }

  const array = {
    append: (path: PathFormPath, item: any) => {
      mutateArray(path, (targetArrayStoreItem: PathFormStoreArray) => {
        // push the new store item to the end of the array
        targetArrayStoreItem.value.push(createStoreItem(item));
      });
    },

    prepend: (path: PathFormPath, item: any) => {
      mutateArray(path, (targetArrayStoreItem: PathFormStoreArray) => {
        // unshift the new store item to the beginning of the array
        targetArrayStoreItem.value.unshift(createStoreItem(item));
      });
    },

    move: (path: PathFormPath, fromIndex: number, toIndex: number) => {
      mutateArray(path, (targetArrayStoreItem: PathFormStoreArray) => {
        // mutate the target array
        arrayMove(targetArrayStoreItem.value, fromIndex, toIndex);
      });
    },

    remove: (path: PathFormPath, index: number) => {
      mutateArray(path, (targetArrayStoreItem: PathFormStoreArray) => {
        // mutate the target array
        arrayRemove(targetArrayStoreItem.value, index);
      });
    },
  };

  return (
    <PathFormStateContext.Provider
      value={{
        state,
        watchers,
        array,
        getValues,
        setValue,
        addError,
        clearError,
        clearErrors,
        setMeta,
        forEachStoreItem,
        validate,
        validateStore,
        isDirty,
        reset,
      }}
    >
      {children}
    </PathFormStateContext.Provider>
  );
};

export function usePathForm() {
  const context = React.useContext(PathFormStateContext) as PathFormStateContext;

  if (!context || context.state === null || context.watchers === null) {
    throw new Error('`usePathForm` must be used within a `PathFormProvider`');
  }

  return context;
}
