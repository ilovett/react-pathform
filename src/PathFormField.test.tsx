import React from 'react';
import { getByLabelText, getByTestId, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { PathFormProvider } from './usePathForm';
import { PathFormField } from './PathFormField';

const TestWrapper: React.FC = ({ children }) => {
  return (
    <PathFormProvider initialRenderValues={{ nested: { items: [{ name: 'Joey Joe Joe Jr. Shabadoo' }] } }}>{children}</PathFormProvider>
  );
};

describe('PathFormField', () => {
  let container: HTMLElement;

  beforeEach(() => {
    ({ container } = render(
      <PathFormField
        path={['nested', 'items', 0, 'name']}
        defaultValue="default"
        render={({ inputProps, meta, renders }) => {
          return (
            <div>
              <label htmlFor="name">Name</label>
              <input id="name" {...inputProps} />
              <pre data-testid="meta">{JSON.stringify(meta)}</pre>
              <pre data-testid="renders">{JSON.stringify(renders)}</pre>
            </div>
          );
        }}
      />,
      { wrapper: TestWrapper }
    ));
  });

  it('renders with the value from the store', async () => {
    expect(getByTestId(container, 'meta')).toMatchInlineSnapshot(`
      <pre
        data-testid="meta"
      >
        {"uuid":"uuid-4","dirty":false,"touched":false,"error":null,"validations":null}
      </pre>
    `);
    expect(getByLabelText(container, 'Name')).toHaveDisplayValue('Joey Joe Joe Jr. Shabadoo');
    expect(getByTestId(container, 'renders')).toMatchInlineSnapshot(`
      <pre
        data-testid="renders"
      >
        0
      </pre>
    `);
  });

  it('user can type and modify the value in the store', async () => {
    userEvent.click(getByLabelText(container, 'Name'));
    userEvent.type(getByLabelText(container, 'Name'), ' new text');

    // new text appended
    expect(getByLabelText(container, 'Name')).toHaveDisplayValue(/ new text$/);

    // rendered 9 times due to typing 9 characters
    // TODO currently its double, 18, until setDirty and setValue trigger one render
    expect(getByTestId(container, 'renders')).toHaveTextContent('18');
  });
});
