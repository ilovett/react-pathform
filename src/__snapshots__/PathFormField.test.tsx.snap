// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PathFormField base renders with the value from the store 1`] = `
<pre
  data-testid="meta"
>
  {"uuid":"uuid-3","dirty":false,"touched":false,"error":null,"validations":null,"defaultValue":"Joey Joe Joe Jr. Shabadoo"}
</pre>
`;

exports[`PathFormField base renders with the value from the store 2`] = `
<pre
  data-testid="renders"
>
  0
</pre>
`;

exports[`PathFormField onChange validations adds an error when another value on the store is invalid 1`] = `
<pre
  data-testid="meta"
>
  {"uuid":"uuid-3","dirty":true,"touched":false,"error":{"type":"custom","message":"Some other item is invalid"},"validations":[{"type":"custom","message":"Some other item is invalid"}],"defaultValue":"Joey Joe Joe Jr. Shabadoo"}
</pre>
`;

exports[`PathFormField onChange validations adds an error when custom validation returns false 1`] = `
<pre
  data-testid="meta"
>
  {"uuid":"uuid-3","dirty":true,"touched":false,"error":{"type":"custom","message":"Not a real name"},"validations":[{"type":"custom","message":"Not a real name"}],"defaultValue":"Joey Joe Joe Jr. Shabadoo"}
</pre>
`;

exports[`PathFormField onChange validations clears the error when it's fixed 1`] = `
<pre
  data-testid="meta"
>
  {"uuid":"uuid-3","dirty":true,"touched":false,"error":{"type":"required","message":"Field is required"},"validations":[{"type":"required","message":"Field is required"}],"defaultValue":"Joey Joe Joe Jr. Shabadoo"}
</pre>
`;

exports[`PathFormField onChange validations clears the error when it's fixed 2`] = `
<pre
  data-testid="meta"
>
  {"uuid":"uuid-3","dirty":true,"touched":false,"error":null,"validations":[{"type":"required","message":"Field is required"}],"defaultValue":"Joey Joe Joe Jr. Shabadoo"}
</pre>
`;

exports[`PathFormField onSubmit validations adds an error when another value on the store is invalid 1`] = `
<pre
  data-testid="meta"
>
  {"uuid":"uuid-3","dirty":true,"touched":true,"error":{"type":"custom","message":"Some other item is invalid"},"validations":[{"type":"custom","message":"Some other item is invalid"}],"defaultValue":"Joey Joe Joe Jr. Shabadoo"}
</pre>
`;

exports[`PathFormField onSubmit validations adds an error when custom validation returns false 1`] = `
<pre
  data-testid="meta"
>
  {"uuid":"uuid-3","dirty":true,"touched":true,"error":{"type":"custom","message":"Not a real name"},"validations":[{"type":"custom","message":"Not a real name"}],"defaultValue":"Joey Joe Joe Jr. Shabadoo"}
</pre>
`;
